File: TreeFromInPre.cpp
// Time Complexity : O(N) where N is the number of elements in the array
// Space Complexity : O(H) where H is the height of the tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nope
// Problem Link: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
// Submission Link: https://leetcode.com/submissions/detail/537349141/

class Solution {
    int idx = 0;
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> indices;
        
        for(int i = 0; i < inorder.size(); i++)
            indices[inorder[i]] = i;
        
        return helper(preorder, indices, 0, inorder.size() - 1);
    }
    
    TreeNode* helper(vector<int>& preorder, unordered_map<int, int>& indices, int low, int high){
        if(low > high)
            return NULL;
        
        int val = preorder[idx++];
        TreeNode* root = new TreeNode(val);
        root -> left = helper(preorder, indices, low, indices[val] - 1);
        root -> right = helper(preorder, indices, indices[val] + 1, high);
        
        return root;
    }
};

File: Tree_from_In_Pre.java
// Time Complexity : O(n) where 'n' is the number of nodes.
// Space Complexity : O(n) where 'n' is the number of nodes.
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*Approach:
store indixed of inorder traversal in a map
start iterating preorder, for the val in pre order, the index at which that value occurs is the root
we do recursive calls to call the helper function on left hald and right half of that inorder array (choosing a value from preorder evrytime)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Tree_from_In_Pre {
    int idx;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < inorder.length ; i++){
                map.put(inorder[i],i);
        }
        return helper(map,0,inorder.length - 1, preorder );
    }
    private TreeNode helper(HashMap<Integer, Integer> map, int start, int end, int[] preorder){
        //base
        if(start > end ) return null;
        int rootVal = preorder[idx];
        idx++;
        int rootIdx = map.get(rootVal);
        TreeNode root = new TreeNode(rootVal);
        root.left = helper(map, start, rootIdx - 1, preorder);
        root.right = helper(map, rootIdx+1,end, preorder);
        return root;
    }
}